File: dynamic_programming/BOJ11066.cpp
#define _CRT_SECURE_NO_WARNINGS

#include <stdio.h>

#define MAX_K 500
#define INF (int)2e9

int T;
int K;
// 파일의 크기 목록
// 1부터 시작
int inputList[MAX_K + 1] = { 0, };
// 부분합
// 1부터 시작
int prefixSum[MAX_K + 1] = { 0, };
// dp[a][b] : a~b번째 장을 합치는데 드는 최소 비용
// 1부터 시작
int dp[MAX_K + 1][MAX_K + 1] = { 0, };

int Min(const int a, const int b)
{
	return ((a < b) ? a : b);
}

void solve(void)
{
	scanf("%d", &T);

	for (register int testCase = 1; testCase <= T; testCase++)
	{
		scanf("%d", &K);

		for (register int i = 1; i <= K; i++)
		{
			scanf("%d", &inputList[i]);

			prefixSum[i] = prefixSum[i - 1] + inputList[i];
		}

		for (register int dist = 1; dist <= K; dist++)
		{
			for (register int start = 1; start + dist <= K; start++)
			{
				int end = start + dist;

				dp[start][end] = INF;

				for (register int mid = start; mid < end; mid++)
				{
					dp[start][end] = Min(dp[start][end], dp[start][mid] + dp[mid + 1][end] + prefixSum[end] - prefixSum[start - 1]);
				}
			}
		}

		printf("%d\n", dp[1][K]);
	}

	return;
}

int main(void)
{
	// freopen("input1.txt", "r", stdin);

	solve();

	return 0;
}

File: dynamic_programming/BOJ11066.py
import sys;

# sys.stdin = open("input1.txt", 'r');

T = int(sys.stdin.readline());

for testCase in range(T):
    fileSizeList = [0];

    K = int(sys.stdin.readline());
    fileSizeList.extend(list(map(int, sys.stdin.readline().split())));

    prefixSumList = [0 for _ in range(K + 1)];
    dp = [[0 for _ in range(K + 1)] for _ in range(K + 1)];

    for curIdx in range(1, K + 1):
        prefixSumList[curIdx] = prefixSumList[curIdx - 1] + fileSizeList[curIdx];

    # print(K);
    # print(fileSizeList);
    # print(prefixSumList);

    for i in range(2, K + 1):
        for j in range(1, K - i + 2):
            dp[j][j + i - 1] = min([dp[j][j + _] + dp[j + _ + 1][j + i - 1] for _ in range(i - 1)]) + (prefixSumList[j + i - 1] - prefixSumList[j - 1]);

    print(dp[1][K]);
